### ISSUE TITLE
applyImpulse not working on dynamically generated entity?

### ISSUE THREAD

**Issue by davatron5000** [donmccurdy/aframe-physics-system#46]
I've been trying for a week now to piece together a physics-based projectile system. I pieced together a few examples from here on how to get `applyImpulse` to work on a dynamically generated entity.

Reduced test case: https://codepen.io/davatron5000/pen/XgQzEK/?editors=0010

- Using the undocumented `body-loaded` event.
- Copypasta of the working [ForcePush](https://github.com/donmccurdy/aframe-extras/blob/master/examples/force-push/index.html#L77-L88) example.
- Objects have a body, hit the ground, and stack up because they have no foward momentum.
- Don mentioned in Slack that this might need a `setTimeout` but when that's applied, objects hit the floor THEN have the impulse applied.

Not sure if there's a better way to do this or if something is broken here. Filing an issue per @donmccurdy recommendation in Slack.

**Comment by donmccurdy** [comment]
Here's an updated codepen: https://codepen.io/donmccurdy/pen/eRavEv?editors=1010

And annotated code:

```js
const scene = document.querySelector('a-scene');
const camera = document.querySelector('[camera]');

// Use an origin point behind the head, not at the head, so
// there's a useful vector between the origin and the projectile.
const attachment = document.querySelector('#attachment');

function spawnBullet() {
    let entity = document.createElement('a-sphere');
    let impulseAmount = 100;
    
    entity.setAttribute('radius', 0.1);
    // Set initial position of projectile to that of the camera.
    entity.setAttribute('position', camera.getAttribute('position'));
    entity.setAttribute('color', '#00FFCC');
    entity.setAttribute('shader', 'flat');
    entity.setAttribute('mass', 1);
    
    // Append projectile to the scene, not to the camera, to
    // avoid all sorts of complications.
    scene.appendChild(entity);

    entity.setAttribute('dynamic-body', true);  
    
    entity.addEventListener('body-loaded', function(){
      // Can't apply forces during the same tick that attaches the body, because
      // it hasn't been fully synced to the physics sim. (bug)
      setTimeout(function () {
        let pStart = new CANNON.Vec3();
        pStart.copy(attachment.object3D.getWorldPosition());
        let force = entity.body.position.vsub(pStart);
        force.normalize(); 
        force.scale(impulseAmount, force);
        entity.body.applyImpulse(force, entity.body.position);
      }, 0);
    });
}

// Scene may have loaded already. If so, initialize.
if (scene.hasLoaded) init();
else scene.addEventListener('loaded', init);

function init () {
  // any code that appends things to the scene
  scene.addEventListener('click', spawnBullet);
}

```

**Comment by davatron5000** [comment]
Thanks so much, @donmccurdy! Removed my blocking issue. 

- I suspected something like `attachment` was needed to position the spawn away from the force initiator. 
- The `setTimeout( function(){}, 0)` helped.
- The `attachment.object3D.getWorldPosition()` was also critical, That's kind of a Three.js feature I didn't really know about yet. 

Again thanks! I was blocked on this for about two weeks. I'll have an A-Frame-based game out soon. Let me know if I can buy you beers. :)
